pkg/cache/multi_namespace_cache.go

```diff
@@ -166,6 +174,18 @@ func (c *multiNamespaceCache) List(ctx context.Context, list client.ObjectList,
 		if err != nil {
 			return err
 		}
+		if limitSet {
+			// decrement Limit by the number of items
+			// fetched from the current namespace.
+			listOpts.Limit -= int64(len(items))
+			// if the Limit is exceeded while
+			// reading from the current namespace,
+			// then keep only the number of items
+			// needed to not overshoot the Limit.
+			if listOpts.Limit < 0 {
```

Comment:
Since each individual cache will respect the limit, this shouldn't be possible?

Reply:
My understanding was that, since `opts` is being passed into the `List`, if the original `Limit` was `3`, and if we go with a similar configuration as in the test, from the first namespace it would read out `2` items (`limitOpts.Limit` now being `1`), in the `List` of the next namepsace, the passed in `ListOptions` still would have a `Limit` of `3`, it would read out `2` items from the second namespace and the `limitOpts.Limit` would now be `-1` and the total `List`ed items would be `4`, which overshoots the original `Limit`

wdyt?

Reply:
You are forgetting that we adjust `listOpts.Limit` after every time we read something: `listOpts.Limit -= int64(len(items))`

Reply:
If I understand correctly, the decremented `listOpts.Limit` won't really be reflected in `opts`?
https://github.com/kubernetes-sigs/controller-runtime/blob/3b25aa6a620cd80323abc0f31613bb1ab2014d08/pkg/cache/multi_namespace_cache.go#L192

Reply:
Well, replace `opts...` with `listOpts`